src/controllers/notifications.js

```diff
@@ -71,6 +73,16 @@ const UnacceptableExpiryError = require('../errors/unacceptable-expiry-error')
 exports.post = function * postNotification () {
   const notification = yield requestUtil.validateBody(this, 'Notification')
   log.debug('Got notification: ' + JSON.stringify(notification))
+
+  const verifySignature = this.config.get('notifications.must_verify')
+  if (verifySignature) {
+    const result = model.verifySignature(notification, this.config)
+    if (!result.valid) {
+      log.warn('Signature verification failed: ' + result.error)
+      throw new UnprocessableEntityError('Notification failed signature verification')
```

Comment:
422 seems like the correct code to return. (401 would be more specific, but we're not using a standard HTTP authentication method, so it doesn't work in our case.)

Regarding retry behavior: Generally, we retry on 4xx even though it goes against HTTP. We do this because we would like to be conservative. We should have reasonable backoff and intelligent scheduling to avoid hammering a server to death. But the protocol very heavily relies on reliable delivery, so giving up on a notification should be a very last resort.

> I haven't seen a proxy server or load balancer return a 4XX when the upstream server is at fault, that's usually what 502, 503, and 504 are reserved for.

It commonly happens when a load balancer is pointing at the wrong server, so the server is responding with 404 or 403 because it doesn't know the URL. It could also be a misconfigured component, for instance the connector will respond 422 if it doesn't recognize the asset pair, which would happen if it is temporarily loading the wrong config.

We could stop retrying if we see InvalidKeyError, but I think that's the exact opposite of what we should do. The scenario I'm imagining is that a ledger somehow has a misconfigured key. Their notifications start failing, so they get notified. They go in and fix the configuration and restart the server. If we make the behavior non-retry, they now have to somehow reset the state of the failed notifications. If we make the behavior retry, they don't have to do anything, things will just work themselves out.